### ISSUE TITLE
Why is it interesting to check alg and enc with header checker when JWS/JWE loaders already checks it?

### ISSUE THREAD

**Issue by ziming** [web-token/jwt-framework#196]
In the docs it says this:

> even if the JWS and JWE Loaders will check the alg/enc header parameters, it is interesting to check them through this manager. 

Why is it interesting to check them with header checker? Just for curiosity.

**Comment by Spomky** [comment]
Normally, the token loading process is very similar to [the one described in the documentation](https://web-token.spomky-labs.com/security-recommendations#loading-process).

The `alg`/`enc` check is a very trival operation to perform (call `inArray` PHP function).

If you don’t check the header, the process will continue to the next step that is the verify the signature (JWS) or the decryption the token (JWE).
The library will first get the algorithm and, if it is in you manager, will check the keys in your keyset are suitable for it and try to verify the signature.
This process is much more longer than the simple `inArray` call, especially if you have lot of keys in your keyset (because you allow several algorithms).

With the header checks, you can reject invalid token earlier and avoid unecessary operations.

**Comment by ziming** [comment]
Ah okay @Spomky 

So if I got you correctly, although the JWS/JWE loader also checks for the alg and enc headers. It is better to first use HeaderManager to check it so that if it is the wrong alg and enc, we can reject it early instead of running the full JWS verification or JWE decryption process?

**Comment by Spomky** [comment]
> we can reject it early instead of running the full JWS verification or JWE decryption process

That’s it!

**Comment by ziming** [comment]
But wouldn't most requests be legit ones and hence this header checker actually slow things down in practice?

**Comment by Spomky** [comment]
I prepared some tests to check if it actually slow down the process or not.
In the attached package, there are 2 tests:
* The first one with the `alg` header parameter checker,
* The second one without any header checker. The checker manager is still called even if empty to be is the same configuration (it also checks the `crit`  header).

I also used a key set with 6 keys (common usage if key rotation + different key types).
On my platform, it takes approx 20ms first the first case (with checker) to reject the token and 25ms for the second one (without checker) i.e. approx. +25%.

Even if 5ms is not really significant, it may have an impact for large systems dealing with a high number of requests to manage. I agree with you most requests are legit and it will not change anything in that case. However this check costs nothing but may be useful it case of an attack.
That choice clearly depends on the application context.

I hope it helps.

Note: to run that test, unzip the package and run the following commands:

```sh
composer install
vendor/bin/phpbench run --store
vendor/bin/phpbench report --report=full --output=md --uuid=latest
```

[benchmark.zip](https://github.com/web-token/jwt-framework/files/3558565/benchmark.zip)
